Source/UI/RawMessageView.swift

```diff
@@ -0,0 +1,147 @@
+//
+//  RawMessageView.swift
+//  Planetary
+//
+//  Created by Martin Dutra on 19/9/22.
+//  Copyright © 2022 Verse Communications Inc. All rights reserved.
+//
+
+import SwiftUI
+
+/// A view model for the RawMessageView
+@MainActor protocol RawMessageViewModel: ObservableObject {
+
+    /// The raw message to display in screen
+    var rawMessage: String? { get }
+
+    /// A loading message that should be displayed when it is not nil
+    var loadingMessage: String? { get }
+
+    /// An error message that should be displayed when it is not nil
+    var errorMessage: String? { get }
+
+    /// Called when the user dismisses the shown error message. Should clear `errorMessage`.
+    func didDismissError()
+
+    /// Called when the user taps on the Cancel button
+    func didDismiss()
+}
+
+struct RawMessageView<ViewModel>: View where ViewModel: RawMessageViewModel {
+    @ObservedObject var viewModel: ViewModel
+
+    /// A loading overlay that displays the `loadingMessage` from the view model.
+    private var loadingIndicator: some View {
+        VStack {
+            Spacer()
+            if showProgress, let loadingMessage = viewModel.loadingMessage {
+                VStack {
+                    PeerConnectionAnimationView(peerCount: 5)
+                    SwiftUI.Text(loadingMessage)
+                        .foregroundColor(Color("mainText"))
+                }
+                .padding(16)
+                .cornerRadius(8)
+                .background(Color("cardBackground").cornerRadius(8))
+            } else {
+                EmptyView()
+            }
+            Spacer()
+        }
+    }
+
+    private var showAlert: Binding<Bool> {
+        Binding {
+            viewModel.errorMessage != nil
+        } set: { _ in
+            viewModel.didDismissError()
+        }
+    }
+
+    private var showProgress: Bool {
+        viewModel.loadingMessage?.isEmpty == false
+    }
+    
+    var body: some View {
+        ZStack {
+            if let source = viewModel.rawMessage {
+                SelectableText(text: source)
```

Comment:
The `SelectableText` thing seems to have worked out nicely! I bet we will use this a lot in the future.